### PR TITLE
[Merged by Bors] - fix: alias FVarIds for `change` and `clear_value`

### PR DESCRIPTION
**Pull request by kmill**
The tactics `change ... at` and `clear_value` do a `revert`/`intro` pattern, but the `FVarId`s before and after `revert`/`intro` weren't given an `FVarAliasInfo` record in the info tree. This could cause the unused variable linter to not see that variables were being used.

---

- [x] depends on: #2916

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2916
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by digama0** [comment]
Whoa, this is not ok. FVarIds are used in caches, it is a fundamental property of the architecture that FVarIds are never reused. To solve issues with revert/intro you should be using `FVarSubst` to keep track of the renaming map and put it in the info tree so that the linter can follow along.

**Comment by kmill** [comment]
@digama0 Could you be more specific about what isn't ok here? `FVarId`s are definitely reused in Lean 4, for example by `dsimp`:
```
import Lean

open Lean Elab Elab.Tactic Meta

elab stx:"list_fvars" : tactic => do
  let g ← getMainGoal
  let lctx := (← g.getDecl).lctx
  logInfoAt stx s!"{lctx.getFVarIds.map (·.name)}"

def Inhabited' α := Inhabited α

@[simp]
theorem Inhabited'_def : Inhabited' α = Inhabited α := rfl

example (x : id Inhabited' (id α)) : Inhabited α := by
  list_fvars -- #[_uniq.2120, _uniq.2121, _uniq.2122]
  dsimp at x
  list_fvars -- #[_uniq.2120, _uniq.2121, _uniq.2122]
  assumption
```
This seems sketchy to me though for other reasons, since (as you can check) `dsimp` doesn't register any type hints when it rewrites local hypotheses, unlike doing `dsimp` on the goal.

Assuming `dsimp` is ok, then is it that clearing the value of a local definition not ok? I'm happy to switch to using `FVarSubst` (I tried asking about whether there was some way to link FVarIds [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/clear_value/near/342363763)), but it would be nice to know some concrete example where clearing a value breaks things.

**Comment by digama0** [comment]
I think Leo and Gabriel are the best ones to ask about what concretely goes wrong when local constant "unique names" aren't unique. I would just recommend avoiding anything like it, and if `dsimp at x` is doing that then there are probably ways to demonstrate incorrect behavior using it as well.

**Comment by kmill** [comment]
I haven't been able to see how `FVarSubst` is useful for recording these substitutions in the info tree (sure `substCore` maintains one, but it seems like it's never used in relation to info trees). It's easy enough to insert the aliases pair-by-pair using [@gebner's suggestion](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.60dsimp.20.2E.2E.2E.20at.60.20reusing.20FVarIds.2C.20instance.20cache.20not.20updated/near/343272304) though.

File: Mathlib/Tactic/Basic.lean [hunk 1]
```diff
@@ -31,6 +31,38 @@ syntax (name := lemma)
...
+      if checkDefEq then
+        unless ← isDefEq typeNew typeOld do
+          throwTacticEx `changeLocalDecl mvarId
+            m!"given type{indentExpr typeNew}\nis not definitionally equal to{indentExpr typeOld}"
+    let finalize (targetNew : Expr) : MetaM MVarId := do
+      let mvarId ← mvarId.replaceTargetDefEq targetNew
+      let (ys, mvarId) ← mvarId.introNP xs.size
+      mvarId.withContext do
+        for x in xs, y in ys do
+          pushInfoLeaf (.ofFVarAliasInfo { id := y, baseId := x, userName := ← y.getUserName })
```
Comment (gebner) [inline Mathlib/Tactic/Basic.lean:59]: I think it would be great to extract this snippet into a separate function (it already occurs twice in this PR)

**Review by gebner — COMMENTED** [review]
otherwise lgtm

File: Mathlib/Tactic/Basic.lean (hunk 1 above)
Comment (kmill) [reply Mathlib/Tactic/Basic.lean]: I made a little API for the revert/intro pattern that's hopefully useful beyond `change` and `clear_value`, and which is hopefully not over-engineered.

I extracted the snippet too, but it's not used in this API.

**Pull request closed without merging**